File: algorithms/2_medium/M/03976_maximum_subarray_sum_after_multiplier/solution.c
struct data {
    int64_t by;
    int64_t by_mul;
    int64_t by_div;
};

static inline int64_t max_int64(const int64_t a, const int64_t b)
{
    return a > b ? a : b;
}

static inline struct data* init_max(const int* nums, const int len, const int64_t k)
{
    struct data* max = malloc(((size_t)len + 1) * sizeof(*max));
    assert(max != NULL);
    max[len] = (struct data){ by: 0, by_mul: 0, by_div: 0};
    int64_t by = 0;
    int64_t by_mul = 0;
    int64_t by_div = 0;
    for (int i = len - 1; i >= 0; i -= 1) {
        const int64_t num = nums[i];
        by_mul = max_int64(0, max_int64(by_mul, by) + num * k);
        by_div = max_int64(0, max_int64(by_div, by) + num / k);
        by = max_int64(0, by + num);
        max[i] = (struct data){ by: by, by_mul: by_mul, by_div: by_div};
    }

    return max;
}

long long maxSubarraySum(const int* nums, const int len, const int k)
{
    assert(k >= 1);
    const struct data* max = init_max(nums, len, k);
    int64_t by_l = 0;
    int64_t by_mul_l = 0;
    int64_t by_div_l = 0;
    int64_t res = INT64_MIN;

    for (int i = 0; i < len; i += 1) {
        const int64_t num = nums[i];
        int64_t by_r = max[i + 1].by;
        int64_t by_mul_r = max[i + 1].by_mul;
        int64_t by_div_r = max[i + 1].by_div;
        by_mul_l = max_int64(0 ,by_mul_l) + num * k;
        by_div_l = max_int64(0, by_div_l) + num / k;
        by_l = max_int64(0, by_l) + num;

        res = max_int64(res, by_mul_l + by_r);
        res = max_int64(res, by_div_l + by_r);
        res = max_int64(res, by_l + by_mul_r);
        res = max_int64(res, by_l + by_div_r);
    }

/* cleanup */
    free(max);
    return res;
}

File: algorithms/2_medium/M/03976_maximum_subarray_sum_after_multiplier/solution.rs
impl Solution {
    #[inline(always)]
    fn init_max(nums: &[i32], k: i64) -> Vec<(i64, i64, i64)> {
        let len = nums.len();
        let mut max = vec![(0_i64, 0_i64, 0_i64); len + 1];
        let (mut by, mut by_mul, mut by_div) = (0_i64, 0_i64, 0_i64);
        for i in (0..len).rev() {
            let num = nums[i] as i64;
            by_mul = (by_mul.max(by) + num * k).max(0);
            by_div = (by_div.max(by) + num / k).max(0);
            by = (by + num).max(0);
            max[i] = (by, by_mul, by_div);
        }

        max
    }

    pub fn max_subarray_sum(nums: Vec<i32>, k: i32) -> i64 {
        assert!(k >= 1);
        let k = k as i64;
        let len = nums.len();
        let max = Self::init_max(&nums, k);
        let (mut by_l, mut by_mul_l, mut by_div_l) = (0_i64, 0_i64, 0_i64);
        let mut res = i64::MIN;

        for (i, num_i32) in nums.into_iter().enumerate() {
            let (by_r, by_mul_r, by_div_r) = max[i + 1];
            let num = num_i32 as i64;
            by_mul_l = by_mul_l.max(0) + num * k;
            by_div_l = by_div_l.max(0) + num / k;
            by_l = by_l.max(0) + num;
            res = res.max(by_mul_l + by_r).max(by_div_l + by_r).max(by_l + by_mul_r).max(by_l + by_div_r);
        }
        
        res
    }
}
